File: zig/src/json.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const JsonValue = union(enum) {
    null,
    bool: bool,
    number: f64,
    string: []const u8,
    array: []JsonValue,
    object: std.StringArrayHashMap(JsonValue),
};

pub const Parser = struct {
    text: []const u8,
    pos: usize,
    allocator: Allocator,

    pub fn init(allocator: Allocator, text: []const u8) Parser {
        return .{
            .text = text,
            .pos = 0,
            .allocator = allocator,
        };
    }

    fn skipWs(self: *Parser) void {
        while (self.pos < self.text.len) {
            switch (self.text[self.pos]) {
                ' ', '\n', '\r', '\t' => self.pos += 1,
                else => break,
            }
        }
    }

    fn peek(self: *Parser, expected: u8) bool {
        return self.pos < self.text.len and self.text[self.pos] == expected;
    }

    fn expect(self: *Parser, expected: u8) !void {
        self.skipWs();
        if (!self.peek(expected)) return error.ExpectedChar;
        self.pos += 1;
    }

    pub fn parseValue(self: *Parser) anyerror!JsonValue {
        self.skipWs();
        if (self.pos >= self.text.len) return error.UnexpectedEnd;
        return switch (self.text[self.pos]) {
            '{' => try self.parseObject(),
            '[' => try self.parseArray(),
            '"' => .{ .string = try self.parseString() },
            't', 'f' => .{ .bool = try self.parseBool() },
            'n' => blk: {
                try self.parseNull();
                break :blk .null;
            },
            '-', '0'...'9' => .{ .number = try self.parseNumber() },
            else => error.UnexpectedCharacter,
        };
    }

    fn parseObject(self: *Parser) anyerror!JsonValue {
        try self.expect('{');
        var object = std.StringArrayHashMap(JsonValue).init(self.allocator);
        errdefer object.deinit();

        self.skipWs();
        if (self.peek('}')) {
            self.pos += 1;
            return .{ .object = object };
        }

        while (true) {
            self.skipWs();
            const key = try self.parseString();
            self.skipWs();
            try self.expect(':');
            const value = try self.parseValue();
            try object.put(key, value);
            self.skipWs();
            if (self.peek('}')) {
                self.pos += 1;
                return .{ .object = object };
            }
            try self.expect(',');
        }
    }

    fn parseArray(self: *Parser) anyerror!JsonValue {
        try self.expect('[');
        var array = std.ArrayList(JsonValue).init(self.allocator);
        errdefer array.deinit();

        self.skipWs();
        if (self.peek(']')) {
            self.pos += 1;
            return .{ .array = try array.toOwnedSlice() };
        }

        while (true) {
            try array.append(try self.parseValue());
            self.skipWs();
            if (self.peek(']')) {
                self.pos += 1;
                return .{ .array = try array.toOwnedSlice() };
            }
            try self.expect(',');
        }
    }

    fn parseString(self: *Parser) ![]const u8 {
        try self.expect('"');
        var out = std.ArrayList(u8).init(self.allocator);
        errdefer out.deinit();

        while (self.pos < self.text.len) {
            const c = self.text[self.pos];
            self.pos += 1;
            if (c == '"') {
                return try out.toOwnedSlice();
            }
            if (c == '\\') {
                if (self.pos >= self.text.len) return error.UnterminatedEscape;
                const esc = self.text[self.pos];
                self.pos += 1;
                switch (esc) {
                    '"', '\\', '/' => try out.append(esc),
                    'b' => try out.append(0x08),
                    'f' => try out.append(0x0c),
                    'n' => try out.append('\n'),
                    'r' => try out.append('\r'),
                    't' => try out.append('\t'),
                    'u' => {
                        if (self.pos + 4 > self.text.len) return error.InvalidUnicodeEscape;
                        const hex = self.text[self.pos .. self.pos + 4];
                        const code = try std.fmt.parseInt(u21, hex, 16);
                        var buf: [4]u8 = undefined;
                        const len = try std.unicode.utf8Encode(code, &buf);
                        try out.appendSlice(buf[0..len]);
                        self.pos += 4;
                    },
                    else => return error.InvalidEscape,
                }
            } else {
                try out.append(c);
            }
        }
        return error.UnterminatedString;
    }

    fn parseNumber(self: *Parser) !f64 {
        const start = self.pos;
        if (self.peek('-')) self.pos += 1;
        while (self.pos < self.text.len and self.text[self.pos] >= '0' and self.text[self.pos] <= '9') {
            self.pos += 1;
        }
        if (self.peek('.')) {
            self.pos += 1;
            while (self.pos < self.text.len and self.text[self.pos] >= '0' and self.text[self.pos] <= '9') {
                self.pos += 1;
            }
        }
        if (self.pos < self.text.len and (self.text[self.pos] == 'e' or self.text[self.pos] == 'E')) {
            self.pos += 1;
            if (self.peek('+') or self.peek('-')) self.pos += 1;
            while (self.pos < self.text.len and self.text[self.pos] >= '0' and self.text[self.pos] <= '9') {
                self.pos += 1;
            }
        }
        return try std.fmt.parseFloat(f64, self.text[start..self.pos]);
    }

    fn parseBool(self: *Parser) !bool {
        if (std.mem.startsWith(u8, self.text[self.pos..], "true")) {
            self.pos += 4;
            return true;
        }
        if (std.mem.startsWith(u8, self.text[self.pos..], "false")) {
            self.pos += 5;
            return false;
        }
        return error.InvalidBoolean;
    }

    fn parseNull(self: *Parser) !void {
        if (std.mem.startsWith(u8, self.text[self.pos..], "null")) {
            self.pos += 4;
            return;
        }
        return error.InvalidNull;
    }

    pub fn finish(self: *Parser) !void {
        self.skipWs();
        if (self.pos != self.text.len) return error.TrailingContent;
    }
};

pub fn parse(allocator: Allocator, text: []const u8) !JsonValue {
    var parser = Parser.init(allocator, text);
    const value = try parser.parseValue();
    try parser.finish();
    return value;
}

File: rust/src/json.rs
use std::collections::BTreeMap;

#[derive(Clone, Debug)]
pub enum JsonValue {
    Null,
    Bool(bool),
    Number(f64),
    String(String),
    Array(Vec<JsonValue>),
    Object(BTreeMap<String, JsonValue>),
}

pub struct Parser<'a> {
    text: &'a str,
    pos: usize,
}

impl<'a> Parser<'a> {
    pub fn new(text: &'a str) -> Self {
        Self { text, pos: 0 }
    }

    fn skip_ws(&mut self) {
        while let Some(c) = self.text[self.pos..].chars().next() {
            if c == ' ' || c == '\n' || c == '\r' || c == '\t' {
                self.pos += c.len_utf8();
            } else {
                break;
            }
        }
    }

    fn peek(&self, expected: char) -> bool {
        self.text[self.pos..].starts_with(expected)
    }

    fn expect(&mut self, expected: char) -> Result<(), String> {
        self.skip_ws();
        if !self.peek(expected) {
            return Err(format!("Expected '{expected}'"));
        }
        self.pos += expected.len_utf8();
        Ok(())
    }

    pub fn parse_value(&mut self) -> Result<JsonValue, String> {
        self.skip_ws();
        let c = self
            .text
            .chars()
            .nth(self.pos)
            .ok_or_else(|| "Unexpected end of JSON".to_string())?;
        match c {
            '{' => self.parse_object(),
            '[' => self.parse_array(),
            '"' => Ok(JsonValue::String(self.parse_string()?)),
            't' | 'f' => Ok(JsonValue::Bool(self.parse_bool()?)),
            'n' => {
                self.parse_null()?;
                Ok(JsonValue::Null)
            }
            '-' | '0'..='9' => Ok(JsonValue::Number(self.parse_number()?)),
            _ => Err("Unexpected character in JSON".to_string()),
        }
    }

    fn parse_object(&mut self) -> Result<JsonValue, String> {
        self.expect('{')?;
        let mut object = BTreeMap::new();
        self.skip_ws();
        if self.peek('}') {
            self.pos += 1;
            return Ok(JsonValue::Object(object));
        }
        loop {
            self.skip_ws();
            let key = self.parse_string()?;
            self.skip_ws();
            self.expect(':')?;
            let value = self.parse_value()?;
            object.insert(key, value);
            self.skip_ws();
            if self.peek('}') {
                self.pos += 1;
                return Ok(JsonValue::Object(object));
            }
            self.expect(',')?;
        }
    }

    fn parse_array(&mut self) -> Result<JsonValue, String> {
        self.expect('[')?;
        let mut array = Vec::new();
        self.skip_ws();
        if self.peek(']') {
            self.pos += 1;
            return Ok(JsonValue::Array(array));
        }
        loop {
            array.push(self.parse_value()?);
            self.skip_ws();
            if self.peek(']') {
                self.pos += 1;
                return Ok(JsonValue::Array(array));
            }
            self.expect(',')?;
        }
    }

    fn parse_string(&mut self) -> Result<String, String> {
        self.expect('"')?;
        let mut out = String::new();
        while self.pos < self.text.len() {
            let c = self.text[self.pos..].chars().next().unwrap();
            self.pos += c.len_utf8();
            if c == '"' {
                return Ok(out);
            }
            if c == '\\' {
                let esc = self
                    .text
                    .chars()
                    .nth(self.pos)
                    .ok_or_else(|| "Unterminated escape".to_string())?;
                self.pos += esc.len_utf8();
                match esc {
                    '"' | '\\' | '/' => out.push(esc),
                    'b' => out.push('\u{0008}'),
                    'f' => out.push('\u{000c}'),
                    'n' => out.push('\n'),
                    'r' => out.push('\r'),
                    't' => out.push('\t'),
                    'u' => {
                        if self.pos + 4 > self.text.len() {
                            return Err("Invalid unicode escape".to_string());
                        }
                        let hex = &self.text[self.pos..self.pos + 4];
                        let code = u32::from_str_radix(hex, 16)
                            .map_err(|_| "Invalid unicode escape".to_string())?;
                        out.push(char::from_u32(code).unwrap_or('\u{fffd}'));
                        self.pos += 4;
                    }
                    _ => return Err("Invalid escape".to_string()),
                }
            } else {
                out.push(c);
            }
        }
        Err("Unterminated string".to_string())
    }

    fn parse_number(&mut self) -> Result<f64, String> {
        let start = self.pos;
        if self.peek('-') {
            self.pos += 1;
        }
        while self
            .text
            .chars()
            .nth(self.pos)
            .map(|c| c.is_ascii_digit())
            .unwrap_or(false)
        {
            self.pos += 1;
        }
        if self.peek('.') {
            self.pos += 1;
            while self
                .text
                .chars()
                .nth(self.pos)
                .map(|c| c.is_ascii_digit())
                .unwrap_or(false)
            {
                self.pos += 1;
            }
        }
        if matches!(self.text.chars().nth(self.pos), Some('e' | 'E')) {
            self.pos += 1;
            if self.peek('+') || self.peek('-') {
                self.pos += 1;
            }
            while self
                .text
                .chars()
                .nth(self.pos)
                .map(|c| c.is_ascii_digit())
                .unwrap_or(false)
            {
                self.pos += 1;
            }
        }
        self.text[start..self.pos]
            .parse()
            .map_err(|_| "Invalid number".to_string())
    }

    fn parse_bool(&mut self) -> Result<bool, String> {
        if self.text[self.pos..].starts_with("true") {
            self.pos += 4;
            Ok(true)
        } else if self.text[self.pos..].starts_with("false") {
            self.pos += 5;
            Ok(false)
        } else {
            Err("Invalid boolean".to_string())
        }
    }

    fn parse_null(&mut self) -> Result<(), String> {
        if self.text[self.pos..].starts_with("null") {
            self.pos += 4;
            Ok(())
        } else {
            Err("Invalid null".to_string())
        }
    }

    pub fn finish(&mut self) -> Result<(), String> {
        self.skip_ws();
        if self.pos != self.text.len() {
            Err("Unexpected trailing JSON content".to_string())
        } else {
            Ok(())
        }
    }
}

pub fn parse(text: &str) -> Result<JsonValue, String> {
    let mut parser = Parser::new(text);
    let value = parser.parse_value()?;
    parser.finish()?;
    Ok(value)
}
